src/pyclass.rs

```diff
@@ -107,134 +138,173 @@ pub trait PyClass:
-    }
+fn into_raw<T>(vec: Vec<T>) -> *mut c_void {
+    Box::into_raw(vec.into_boxed_slice()) as _
+}
 
-    // weakref support
-    if let Some(weakref_offset) = PyCell::<T>::weakref_offset() {
-        type_object.tp_weaklistoffset = weakref_offset as ffi::Py_ssize_t;
-    }
+pub(crate) fn create_type_object<T>(
+    py: Python,
+    module_name: Option<&str>,
+) -> PyResult<*mut ffi::PyTypeObject>
+where
+    T: PyClass,
+{
+    let mut slots = TypeSlots::default();
 
-    // GC support
-    if let Some(gc) = T::gc_methods() {
-        unsafe { gc.as_ref() }.update_typeobj(type_object);
-    }
+    slots.push(ffi::Py_tp_base, T::BaseType::type_object_raw(py) as _);
+    slots.maybe_push(ffi::Py_tp_doc, tp_doc::<T>()?);
+    slots.maybe_push(ffi::Py_tp_dealloc, tp_dealloc::<T>().map(|v| v as _));
 
-    // descriptor protocol
-    if let Some(descr) = T::descr_methods() {
-        unsafe { descr.as_ref() }.update_typeobj(type_object);
+    let (new, call, methods) = py_class_method_defs::<T>();
+    slots.maybe_push(ffi::Py_tp_new, new.map(|v| v as _));
+    slots.maybe_push(ffi::Py_tp_call, call.map(|v| v as _));
+    // normal methods
+    if !methods.is_empty() {
+        slots.push(ffi::Py_tp_methods, into_raw(methods));
     }
 
-    // iterator methods
-    if let Some(iter) = T::iter_methods() {
-        unsafe { iter.as_ref() }.update_typeobj(type_object);
+    // properties
+    let props = py_class_properties::<T>();
+    if !props.is_empty() {
+        slots.push(ffi::Py_tp_getset, into_raw(props));
     }
 
-    // nb_bool is a part of PyObjectProtocol, but should be placed under tp_as_number
-    let mut nb_bool = None;
     // basic methods
     if let Some(basic) = T::basic_methods() {
-        unsafe { basic.as_ref() }.update_typeobj(type_object);
-        nb_bool = unsafe { basic.as_ref() }.nb_bool;
+        unsafe { basic.as_ref() }.update_slots(&mut slots);
     }
-
     // number methods
-    type_object.tp_as_number = T::number_methods()
-        .map(|mut p| {
-            unsafe { p.as_mut() }.nb_bool = nb_bool;
-            p.as_ptr()
-        })
-        .unwrap_or_else(|| nb_bool.map_or_else(ptr::null_mut, ffi::PyNumberMethods::from_nb_bool));
+    if let Some(number) = T::number_methods() {
+        unsafe { number.as_ref() }.update_slots(&mut slots);
+    }
+    // iterator methods
+    if let Some(iter) = T::iter_methods() {
+        unsafe { iter.as_ref() }.update_slots(&mut slots);
+    }
     // mapping methods
-    type_object.tp_as_mapping = T::mapping_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
+    if let Some(mapping) = T::mapping_methods() {
+        unsafe { mapping.as_ref() }.update_slots(&mut slots);
+    }
     // sequence methods
-    type_object.tp_as_sequence = T::sequence_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
+    if let Some(sequence) = T::sequence_methods() {
+        unsafe { sequence.as_ref() }.update_slots(&mut slots);
+    }
```

Comment:
I wonder if to clean all of these up we could have a private trait

```
trait PyMethodSlotsDef {
   fn update_slots(slots: &mut TypeSlots)
}
```

and move all the `update_slots` methods onto that trait.

Then we can add a `maybe_push_slots` method to `TypeSlots` so that these can all just be

```
slots.maybe_push_slots(T::sequence_methods())
```

etc.

Reply:
I'm not sure it gets much shorter, to be honest.

Reply:
I'm going to remove these method tables before merging.